### ISSUE TITLE
Add User-Agent header to all requests to the hub

### ISSUE THREAD

**Issue by Kakulukian** [huggingface/huggingface_hub#1018]
Is there a way to add `huggingface_hub`'s custom `User-Agent` to all the requests to the hub? Right now, it seems to be only in files_download. See here
https://github.com/huggingface/huggingface_hub/blob/a1282ab170e6248a893764d5d6759efeaa7a5421/src/huggingface_hub/file_download.py#L317-L354

This would be useful to allow us to have more information about the client versions that use specific API endpoints (e.g: community, commit...).

**Comment by Wauplin** [comment]
Hey @Kakulukian, I have looked into your feature request a bit.
As a disclaimer, I just learnt about this `http_user_agent` and the user agent tracking in general. Maybe I'm missing some context here, let me know if it's the case.

---
That said, my first reaction to "Is there a way to add `huggingface_hub`'s custom User-Agent to all the requests to the hub?" was "uuuurgh, if that means adding `library_name`, `library_version` and `user_agent` to all functions, that will become ugly".

My main concern with having those arguments as public everywhere is that for my understanding they should only matter to librairies contributors (transformers, datasets, skops,...) more than the end-users. I don't see an end-user using those args just to help us track their activity. I'm quite reluctant to the explicit `user_agent` argument because we are quite dependant on if it is used are not.

In general, I think a good design for this would be to have a "HubClient" object in `huggingface_hub` that would be first initialized with token, url endpoint, user-agent,... and then use the "HubClient" to make all the calls. This is what is started with the `HfApi` object but at the moment it's not used everywhere (`cached_download`, `snapshot_download` are stand-alone functions for example). And also we encourage users to use our aliases (example: `create_commit` is an alias of `HfApi().create_commit`) so they don't really instantiate the client themselves. I don't think we will change this in the short-term.

---
### An in-between solution: track versions but implicitely/automatically

An in-between solution I see is to have a tracker on every call to the hub that we handle internally. This is already what we are doing to retrieve pytorch/tensorflow/fastcore/fastai versions. My proposition is to look at the stack trace in Python to track which library led to the call. The big advantage is that we are not dependant users/other librairies to implement it. The drawback is that it allows less flexibility if a user wants to specifically set a library name/library version value.

Here is a short example of how to do that in python:
```py
import inspect

def get_related_packages():
    packages = {}
    frame = inspect.currentframe()
    while frame is not None:
        package_name = frame.f_globals["__package__"]
        frame = frame.f_back

        if package_name == "": # Means Python runpy -> don't care
            continue
        try:
            if package_name not in packages:
                packages[package_name] = importlib_metadata.version(package_name)
        except importlib_metadata.PackageNotFoundError:
            pass # Usually an python internal library -> don't care
            
    return packages
    
{'huggingface_hub': '0.9.0.dev0', 'pluggy': '1.0.0', 'pytest': '7.1.2'}
```

I quickly tried it for a `from_pretrained` call in the tests (hence `pluggy` and `pytest` versions). In addition to this dict, we would still have to add pytorch/tensorflow/fastcore/fastai versions.

What do you think about it ? Would that be enough without polluting too much the statistics ?

**Comment by Wauplin** [comment]
Also, since we are talking about implicit tracking of package names and versions, is there a privacy issue here ?

I am especially thinking about if a company builds an internal library and don't want us to track the name/version of it. With the implementation I suggested it is quite hard to programmatically filter out your package name apart with a global variable or with monkeypatching 🙈 . Can that be a complete no-go ?

**Comment by Wauplin** [comment]
Of course, here I'm only discussing the case where we want to add the user-agent exactly in the same way as in `snapshot_download` with the custom fields `library_name`, `library_version` and `user_agent`. A first implementation can just be to track huggingface_hub/pytorch/tensorflow/fastcore/fastai versions and that's it. I guess this would not be enough for you ?

**Comment by Kakulukian** [comment]
> My main concern with having those arguments as public everywhere is that for my understanding they should only matter to librairies contributors (transformers, datasets, skops,...) more than the end-users. I don't see an end-user using those args just to help us track their activity. I'm quite reluctant to the explicit user_agent argument because we are quite dependant on if it is used are not.
> In general, I think a good design for this would be to have a "HubClient" object in huggingface_hub that would be first initialized with token, url endpoint, user-agent,... and then use the "HubClient" to make all the calls. This is what is started with the HfApi object but at the moment it's not used everywhere (cached_download, snapshot_download are stand-alone functions for example). And also we encourage users to use our aliases (example: create_commit is an alias of HfApi().create_commit) so they don't really instantiate the client themselves. I don't think we will change this in the short-term.

I don't have much opinion on how to do it, it could be something different ofc. For sure, having API client singleton is a way to do it. It seems to me to be the best solution but I'm not a python expert 🙈 I guess it's also possible to monkey patch the function which returns default ua in `requests`

> An in-between solution I see is to have a tracker on every call to the hub that we handle internally. This is already what we are doing to retrieve pytorch/tensorflow/fastcore/fastai versions. My proposition is to look at the stack trace in Python to track which library led to the call. The big advantage is that we are not dependant users/other librairies to implement it. The drawback is that it allows less flexibility if a user wants to specifically set a library name/library version value.
> Of course, here I'm only discussing the case where we want to add the user-agent exactly in the same way as in snapshot_download with the custom fields library_name, library_version and user_agent. A first implementation can just be to track huggingface_hub/pytorch/tensorflow/fastcore/fastai versions and that's it. I guess this would not be enough for you ?


I don't think the goal is to track every library that the end user is using in his project, it's more a way to monitor that we don't introduce a breaking change in the API for many users, for example: we want to change on a key name for a response for an old endpoint, we need to know how many people are using an old version to introduce versioning or not for this endpoint. 
So, yes, it'll be enough to pass these data.

> Also, since we are talking about implicit tracking of package names and versions, is there a privacy issue here?

We can add a telemetry option to disable this behavior, but I think that we have the right to do it according to our [Privacy policy/TOS](https://huggingface.co/privacy), maybe not with all the packages, but at least our (Ping @julien-c on this matter to confirm)



It's not urgent right now, but it's something we might need in the future

**Comment by Wauplin** [comment]
> I don't have much opinion on how to do it, it could be something different ofc. For sure, having API client singleton is a way to do it. It seems to me to be the best solution but I'm not a python expert 🙈 I guess it's also possible to monkey patch the function which returns default ua in requests

If we do it implicitly (no user-added information like `library_name` `library_version` `user_agent`) we will definitely find a way to pass the good user-agent to all requests. My concern is about adding those user arguments everywhere - and will most likely not be used.

> I don't think the goal is to track every library that the end user is using in his project

> maybe not with all the packages, but at least our 

Yes my proposition was more to track every package that led to the call (too from the traceback), not all packages that are installed. In most cases it will be `huggingface_hub` + the library that called `huggingface_hub`.

> maybe not with all the packages, but at least our

If we are talking about tracking just the version of huggingface_hub (+ML libs pytorch/keras/tf/fastai/...) that would be really straightforward to implement.

**Comment by Pierrci** [comment]
> If we are talking about tracking just the version of huggingface_hub (+ML libs pytorch/keras/tf/fastai/...) that would be really straightforward to implement.

Yes! To clarify things, we're solely interested in having this line or something similar for every call to the hub:

https://github.com/huggingface/huggingface_hub/blob/a1282ab170e6248a893764d5d6759efeaa7a5421/src/huggingface_hub/file_download.py#L340

**Comment by Wauplin** [comment]
> Yes! To clarify things, we're solely interested in having this line or something similar for every call to the hub:

Perfect, sorry for the mess 😁 🔥

**Comment by julien-c** [comment]
Agreed!
